File: veritas/training/src/forge_veritas/data/chat_template.py
"""Python port of the CP2 TS chat-template adapter.

Source of truth is `src/lib/veritas/training-format/chat-template.ts` — this
module exists because the training pipeline runs in Python on rented GPU
and can't reach into the TS app at training time. The two implementations
share the same contract; the TS integration test (`integration.ts`) verifies
the wire format the Python exporter produces is parseable by the TS parser.

Why mirror instead of shelling out to Node?
    Two reasons. (1) Performance — millions of episodes will pass through
    this module across SFT + GRPO + DPO. Subprocess overhead of ~50ms per
    episode would add hours to a single export run. (2) No Node on training
    boxes by default — keeping Python self-contained avoids one more thing
    that can go wrong at training time.

Production-grade contract — same as the TS file:
    • reasoning_content is **plain natural-language prose**, no prefix syntax
    • memory recall is a **first-class tool call** (`memory_recall`)
    • tool calls follow the OpenAI shape exactly
    • round-trip is lossless for what Veritas-R1 emits at inference
    • `decide` collapses to `think`, `confidence` is dropped on the wire
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Literal

# Tool-name canon — single source of truth for both training (this file)
# and serving (CP14 vLLM config). MUST match `TOOL_NAMES` in the TS adapter.
TOOL_MEMORY_RECALL = "memory_recall"
TOOL_RETRIEVE = "retrieve"
TOOL_VERIFY = "verify_citation"

ChatRole = Literal["system", "user", "assistant", "tool"]
ThoughtKind = Literal[
    "think", "recall", "retrieve", "verify", "tool-call", "decide", "answer"
]
Mode = Literal["lightning", "reasoning", "deep"]

DEFAULT_SYSTEM_PROMPT = (
    "You are Veritas-R1, Forge's verification-first research assistant. "
    "Ground every answer in the provided claims, episodes, and contradictions. "
    "Prefer abstention to fabrication when evidence is insufficient."
)


# ─────────────────────────────────────────────────────────────────────
#  Message types — match the OpenAI / Qwen3 chat shape exactly
# ─────────────────────────────────────────────────────────────────────


@dataclass(slots=True)
class ToolCall:
    """One assistant tool call. `arguments` is a JSON string per OpenAI shape."""

    id: str
    function_name: str
    arguments: str
    type: str = "function"

    def to_json(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "type": self.type,
            "function": {"name": self.function_name, "arguments": self.arguments},
        }


@dataclass(slots=True)
class ChatMessage:
    """A single chat message. We keep all roles in one dataclass for fast
    construction and uniform JSON encoding — discriminated unions are nice
    in TS but cost real memory in Python at the volumes we're shipping.
    """

    role: ChatRole
    content: str = ""
    reasoning_content: str | None = None
    tool_calls: list[ToolCall] | None = None
    tool_call_id: str | None = None
    name: str | None = None

    def to_json(self) -> dict[str, Any]:
        out: dict[str, Any] = {"role": self.role, "content": self.content}
        if self.reasoning_content:
            out["reasoning_content"] = self.reasoning_content
        if self.tool_calls:
            out["tool_calls"] = [tc.to_json() for tc in self.tool_calls]
        if self.tool_call_id is not None:
            out["tool_call_id"] = self.tool_call_id
        if self.name is not None:
            out["name"] = self.name
        return out


@dataclass(slots=True)
class ThoughtStep:
    """Schema mirror of `ThoughtStep` in `src/lib/veritas/memory/schema.ts`.

    Fields beyond `kind` + `index` are conditional per-kind — this matches the
    TS interface exactly. `confidence` is preserved on the dataclass for
    in-process use (UI annotations, training-time reward shaping) but is
    stripped on the chat-template wire format, same as the TS impl.
    """

    kind: ThoughtKind
    index: int
    text: str | None = None
    recalled_claims: list[str] | None = None
    recalled_episodes: list[str] | None = None
    tool: str | None = None
    tool_input: Any = None
    tool_output: Any = None
    confidence: float | None = None


@dataclass(slots=True)
class ThoughtTrace:
    steps: list[ThoughtStep] = field(default_factory=list)


# ─────────────────────────────────────────────────────────────────────
#  Encoding — ThoughtTrace → ChatMessage[]
# ─────────────────────────────────────────────────────────────────────


def trace_to_chatml(
    trace: ThoughtTrace,
    *,
    user_input: str,
    system_prompt: str | None = None,
    mode: Mode = "reasoning",
) -> list[ChatMessage]:
    """Encode a ThoughtTrace into Qwen3 chat messages.

    Mirrors the TS `traceToChatML` algorithm step-for-step. We deliberately
    keep this implementation byte-for-byte deterministic — same trace, same
    messages, every call. That's the contract CP5 dataset deduplication
    relies on.
    """
    messages: list[ChatMessage] = [
        ChatMessage(role="system", content=system_prompt or DEFAULT_SYSTEM_PROMPT),
        ChatMessage(role="user", content=user_input),
    ]

    strip_reasoning = mode == "lightning"
    reasoning_buf: list[str] = []
    tool_call_counter = 0

    # Walk steps in declared `index` order — schema contract.
    steps = sorted(trace.steps, key=lambda s: s.index)

    def flush_assistant(final_content: str, tool_call: ToolCall | None = None) -> None:
        nonlocal reasoning_buf
        msg = ChatMessage(role="assistant", content=final_content)
        if not strip_reasoning and reasoning_buf:
            # Blank-line separation so multi-paragraph thoughts round-trip.
            msg.reasoning_content = "\n\n".join(reasoning_buf)
        if tool_call is not None:
            msg.tool_calls = [tool_call]
        messages.append(msg)
        reasoning_buf = []

    for step in steps:
        if step.kind in ("think", "decide"):
            if strip_reasoning:
                continue
            text = (step.text or "").strip()
            if text:
                reasoning_buf.append(text)
            continue

        if step.kind == "recall":
            tc_id = f"tc-{tool_call_counter}"
            tool_call_counter += 1
            call = ToolCall(
                id=tc_id,
                function_name=TOOL_MEMORY_RECALL,
                arguments=_jsondumps({"query": step.text or ""}),
            )
            flush_assistant("", call)
            messages.append(
                ChatMessage(
                    role="tool",
                    content=_jsondumps(
                        {
                            "claims": step.recalled_claims or [],
                            "episodes": step.recalled_episodes or [],
                        }
                    ),
                    tool_call_id=tc_id,
                    name=TOOL_MEMORY_RECALL,
                )
            )
            continue

        if step.kind in ("retrieve", "verify", "tool-call"):
            tc_id = f"tc-{tool_call_counter}"
            tool_call_counter += 1
            name = step.tool or _canonical_tool_name(step.kind)
            call = ToolCall(
                id=tc_id,
                function_name=name,
                arguments=_jsondumps(step.tool_input),
            )
            flush_assistant("", call)
            messages.append(
                ChatMessage(
                    role="tool",
                    content=_jsondumps(step.tool_output),
                    tool_call_id=tc_id,
                    name=name,
                )
            )
            continue

        if step.kind == "answer":
            flush_assistant(step.text or "")
            continue

    if reasoning_buf:
        # Trailing reasoning without a closing answer — emit anyway.
        flush_assistant("")

    return messages


def _canonical_tool_name(kind: str) -> str:
    if kind == "retrieve":
        return TOOL_RETRIEVE
    if kind == "verify":
        return TOOL_VERIFY
    return "tool"


# ─────────────────────────────────────────────────────────────────────
#  Decoding — ChatMessage[] → ThoughtTrace
# ─────────────────────────────────────────────────────────────────────


def chatml_to_trace(messages: list[ChatMessage] | list[dict[str, Any]]) -> ThoughtTrace:
    """Reverse direction. Accepts either dataclasses or JSON dicts so callers
    can decode straight from a JSONL line without an extra `from_dict` step.
    """
    msgs = [_coerce_message(m) for m in messages]
    steps: list[ThoughtStep] = []
    idx = 0

    # Index tool-result messages by tool_call_id for O(1) pairing.
    tool_results: dict[str, ChatMessage] = {}
    for m in msgs:
        if m.role == "tool" and m.tool_call_id is not None:
            tool_results[m.tool_call_id] = m

    for m in msgs:
        if m.role != "assistant":
            continue

        # 1) Reasoning prose → one `think` step per blank-line-separated paragraph.
        if m.reasoning_content:
            for raw in _split_paragraphs(m.reasoning_content):
                steps.append(ThoughtStep(kind="think", text=raw, index=idx))
                idx += 1

        # 2) Tool calls — `memory_recall` decodes to `recall`; rest by name.
        if m.tool_calls:
            for tc in m.tool_calls:
                result = tool_results.get(tc.id)
                parsed_output = _jsonloads(result.content) if result else None
                if tc.function_name == TOOL_MEMORY_RECALL:
                    out = parsed_output if isinstance(parsed_output, dict) else {}
                    rc = _string_array(out.get("claims"))
                    re = _string_array(out.get("episodes"))
                    # Recover the query text from the tool call args — preserves
                    # `step.text` on round-trip when a query was supplied.
                    parsed_args = _jsonloads(tc.arguments)
                    query_text: str | None = None
                    if isinstance(parsed_args, dict):
                        q = parsed_args.get("query")
                        if isinstance(q, str) and q:
                            query_text = q
                    step = ThoughtStep(kind="recall", index=idx)
                    if query_text:
                        step.text = query_text
                    if rc:
                        step.recalled_claims = rc
                    if re:
                        step.recalled_episodes = re
                    steps.append(step)
                    idx += 1
                    continue
                steps.append(
                    ThoughtStep(
                        kind=_canonicalise_tool_kind(tc.function_name),
                        index=idx,
                        tool=tc.function_name,
                        tool_input=_jsonloads(tc.arguments),
                        tool_output=parsed_output,
                    )
                )
                idx += 1

        # 3) Final answer text — only if non-empty.
        if m.content:
            steps.append(ThoughtStep(kind="answer", text=m.content, index=idx))
            idx += 1

    return ThoughtTrace(steps=steps)


def _coerce_message(m: ChatMessage | dict[str, Any]) -> ChatMessage:
    if isinstance(m, ChatMessage):
        return m
    tool_calls: list[ToolCall] | None = None
    if m.get("tool_calls"):
        tool_calls = []
        for raw in m["tool_calls"]:
            fn = raw.get("function", {})
            tool_calls.append(
                ToolCall(
                    id=raw.get("id", ""),
                    function_name=fn.get("name", ""),
                    arguments=fn.get("arguments", "{}"),
                    type=raw.get("type", "function"),
                )
            )
    return ChatMessage(
        role=m.get("role", "user"),
        content=m.get("content", ""),
        reasoning_content=m.get("reasoning_content"),
        tool_calls=tool_calls,
        tool_call_id=m.get("tool_call_id"),
        name=m.get("name"),
    )


def _canonicalise_tool_kind(name: str) -> ThoughtKind:
    if name == TOOL_RETRIEVE:
        return "retrieve"
    if name == TOOL_VERIFY:
        return "verify"
    # Heuristics — keep schema kind faithful when the model emits a more
    # specific name (mirrored from the TS adapter for parity).
    low = name.lower()
    if "verify" in low or "doi" in low:
        return "verify"
    if any(t in low for t in ("retrieve", "search", "crossref", "openalex", "arxiv", "pubmed")):
        return "retrieve"
    return "tool-call"


_PARAGRAPH_SPLIT = "\n\n"


def _split_paragraphs(s: str) -> list[str]:
    # Same semantics as the TS `split(/\n\s*\n/)` — split on any blank line,
    # strip whitespace, drop empties.
    parts: list[str] = []
    cur: list[str] = []
    for line in s.split("\n"):
        if line.strip() == "":
            if cur:
                parts.append("\n".join(cur).strip())
                cur = []
        else:
            cur.append(line)
    if cur:
        parts.append("\n".join(cur).strip())
    return [p for p in parts if p]


# ─────────────────────────────────────────────────────────────────────
#  Tiny JSON helpers — never throw, always emit deterministic strings.
#  Determinism matters because identical traces must produce identical
#  messages for dataset dedup at training time.
# ─────────────────────────────────────────────────────────────────────


def _jsondumps(value: Any) -> str:
    if value is None:
        return "{}"
    try:
        # `sort_keys=True` is the cheapest way to make the output canonical
        # without pulling in orjson — same hash across Python builds.
        return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)
    except (TypeError, ValueError):
        return "{}"


def _jsonloads(s: str | None) -> Any:
    if not s:
        return None
    try:
        return json.loads(s)
    except (TypeError, ValueError, json.JSONDecodeError):
        # Surface the raw string rather than dropping data — same as TS.
        return s


def _string_array(v: Any) -> list[str]:
    if not isinstance(v, list):
        return []
    return [x for x in v if isinstance(x, str)]

File: src/lib/veritas/training-format/chat-template.ts
/**
 * Qwen3 chat-template adapter — bidirectional, production-grade.
 *
 * What this file does
 * ───────────────────
 * Translates between Forge's structured `ThoughtTrace.steps` (the schema we
 * persist on every `Episode`) and Qwen3's chat-template message format (the
 * shape `tokenizer.apply_chat_template` consumes during training and emits
 * during inference).
 *
 * Why the seam matters
 * ────────────────────
 * The training pipeline (CP6 SFT, CP9 GRPO, CP11 DPO) packs real Forge
 * episodes into Qwen3 chat messages. The serving path (CP13/14) parses
 * Qwen3 outputs back into ThoughtTrace so they can be persisted on the
 * Episode. **Both directions must be lossless for what the model actually
 * emits at inference** — anything else means the model is trained on a
 * distribution it can't faithfully reproduce, or vice versa.
 *
 * Format choices — and why they match GPT-5 / Claude / o-series exactly
 * ────────────────────────────────────────────────────────────────────
 *   • `reasoning_content` is **plain natural-language prose**. No prefix
 *     syntax, no markup, no embedded structure. This matches:
 *       - OpenAI Responses API: `reasoning.summary` is prose; structure
 *         lives in `tool_calls`.
 *       - Claude content blocks: `thinking` blocks are prose; structure
 *         lives in `tool_use` blocks.
 *       - DeepSeek-R1's `<think>…</think>`: prose.
 *       - Qwen3's pretraining distribution: prose.
 *     A v1 of this file used a `think: …\nrecall: claims=…` prefix syntax —
 *     that was a mistake. It (a) burns 15-25% extra tokens, (b) doesn't match
 *     Qwen3's pretraining distribution, (c) creates a train/serve mismatch
 *     because the model never emits prefix lines at inference.
 *
 *   • Memory recall is a **first-class tool call** (`memory_recall`), not
 *     embedded markup. The model invokes it like any other tool; the tool
 *     response carries the `claims` and `episodes` ids. This is the same
 *     pattern OpenAI uses for `web_search`, Claude uses for `tool_use`, and
 *     it's how Veritas-R1 will actually behave at inference (CP14 wires the
 *     tool definition into vLLM).
 *
 *   • Tool calls follow the OpenAI shape exactly: `tool_calls: [{ id,
 *     type:"function", function: { name, arguments: <JSON-string> } }]`.
 *     vLLM, SGLang, TRL's chat-template loader all read this verbatim.
 *
 * Round-trip contract
 * ───────────────────
 * For every trace `T` that Veritas-R1 produces at inference (kinds:
 * `think | recall | retrieve | verify | tool-call | answer`),
 *      `chatMLToTrace(traceToChatML(T)) == T`
 * modulo step-index renumbering (we always reassign 0..N on output).
 *
 * Schema fields with no inference-time signal — `decide` step kind and
 * `confidence` numbers — are **lossy on the wire format by design**. They
 * are UI annotations for human-authored steps; the model never emits them.
 * `decide` collapses to `think`; `confidence` is dropped. Both fields stay
 * on the schema for internal use (UI, training-data sampling, audit).
 *
 * What this file does NOT do
 * ──────────────────────────
 *   • String-level rendering with `<|im_start|>` markers — that's the
 *     tokenizer's job, performed Python-side via `apply_chat_template`. We
 *     stop at the structured message array.
 *   • Function-call argument validation — the trainer assumes tool inputs
 *     are valid JSON; if a real episode has malformed JSON, the trace
 *     should have been rejected before persistence.
 */

import type { ThoughtStep, ThoughtTrace } from "../memory/schema";

/* ─────────────────────────────────────────────────────────────
 *  Message types — match the OpenAI / Qwen3 chat shape exactly
 * ──────────────────────────────────────────────────────────── */

/** A single tool call emitted by the assistant. `arguments` is a JSON string
 *  to mirror the OpenAI shape exactly — Qwen3 / vLLM / TRL all expect this. */
export interface ChatToolCall {
  id: string;
  type: "function";
  function: {
    name: string;
    arguments: string;
  };
}

/**
 * Roles understood by Qwen3's chat template. `tool` carries function-call
 * results; `assistant` may carry both reasoning and final content.
 */
export type ChatRole = "system" | "user" | "assistant" | "tool";

/** Discriminated message union — keeps each role's required fields obvious. */
export type ChatMessage =
  | SystemMessage
  | UserMessage
  | AssistantMessage
  | ToolMessage;

export interface SystemMessage {
  role: "system";
  content: string;
}

export interface UserMessage {
  role: "user";
  content: string;
}

export interface AssistantMessage {
  role: "assistant";
  /**
   * Final answer text shown to the user. May be empty when this assistant
   * turn only emits tool calls and no answer yet — Qwen3 supports that.
   */
  content: string;
  /**
   * Qwen3-native field for the `<think>…</think>` block. **Plain prose.**
   * Empty or absent when the assistant turn has no reasoning content
   * (lightning mode or pure tool-call turns).
   */
  reasoning_content?: string;
  /**
   * Tool calls the assistant is requesting — surfaces as one or more
   * `<tool_call>{...}</tool_call>` blocks at template-render time.
   */
  tool_calls?: ChatToolCall[];
}

export interface ToolMessage {
  role: "tool";
  /** JSON-stringified tool result. The schema's `toolOutput` is encoded here. */
  content: string;
  /** Matches the `id` of the originating assistant `ChatToolCall`. */
  tool_call_id: string;
  /** Name of the tool, mirrored from the originating call. */
  name?: string;
}

/* ─────────────────────────────────────────────────────────────
 *  Tool-name canon — first-class tools that map onto schema kinds.
 *  The CP14 vLLM serving config wires these tools into the system
 *  prompt; the trainer sees them in CP3/CP6 datasets. Keeping the
 *  names centralised here is the single source of truth.
 * ──────────────────────────────────────────────────────────── */

export const TOOL_NAMES = {
  /** Pull claim / episode ids from project memory. Replaces the v1 `recall:` prefix. */
  memoryRecall: "memory_recall",
  /** External retrieval (Crossref / OpenAlex / arXiv / PubMed). */
  retrieve: "retrieve",
  /** DOI verification — does this paper exist and support the claim? */
  verify: "verify_citation",
} as const;

/* ─────────────────────────────────────────────────────────────
 *  Conversion options
 * ──────────────────────────────────────────────────────────── */

/**
 * Inputs the converter cannot recover from the trace alone.
 *   • `userInput` is the prompt that triggered the trace — stored on
 *     `Episode.input`, not on the trace itself.
 *   • `systemPrompt` is the system preamble used at run time. Defaults to
 *     a minimal verification-first prompt; production wires the same
 *     string Veritas-R1 was trained with (Phase 3 will pin one).
 *   • `mode` selects between Forge UI tiers; `lightning` skips the
 *     reasoning_content emission entirely so chat-mode training data
 *     never carries a stray reasoning block.
 */
export interface TraceToChatMLOptions {
  userInput: string;
  systemPrompt?: string;
  mode?: "lightning" | "reasoning" | "deep";
}

const DEFAULT_SYSTEM_PROMPT =
  "You are Veritas-R1, Forge's verification-first research assistant. " +
  "Ground every answer in the provided claims, episodes, and contradictions. " +
  "Prefer abstention to fabrication when evidence is insufficient.";

/* ─────────────────────────────────────────────────────────────
 *  ThoughtTrace  →  Chat messages
 * ──────────────────────────────────────────────────────────── */

/**
 * Convert a `ThoughtTrace` plus its triggering user input into the structured
 * message array Qwen3's tokenizer consumes.
 *
 * Step-by-step encoding
 * ─────────────────────
 *   • `think` / `decide`  →  contributes a paragraph to `reasoning_content`
 *                            on the next emitted assistant turn. `decide`
 *                            collapses to plain prose (lossy on wire).
 *   • `recall`            →  emits a `memory_recall` tool call + tool-result
 *                            message. Lossless: claim/episode ids round-trip
 *                            through the tool result payload.
 *   • `retrieve`          →  emits a `retrieve` tool call + result.
 *   • `verify`            →  emits a `verify_citation` tool call + result.
 *   • `tool-call`         →  emits a generic tool call using `step.tool` as
 *                            the function name + result.
 *   • `answer`            →  flushes the buffered reasoning into the open
 *                            assistant turn's `reasoning_content` and writes
 *                            its text into `content`.
 *
 * Tool-call ids are deterministic (`tc-0`, `tc-1`, …) so identical traces
 * produce identical message arrays — required for dataset deduplication at
 * training time (CP5).
 */
export function traceToChatML(
  trace: ThoughtTrace,
  opts: TraceToChatMLOptions,
): ChatMessage[] {
  const messages: ChatMessage[] = [
    { role: "system", content: opts.systemPrompt ?? DEFAULT_SYSTEM_PROMPT },
    { role: "user", content: opts.userInput },
  ];

  // In `lightning` mode we strip everything except the answer — chat-mode
  // training data must never carry stray reasoning.
  const stripReasoning = opts.mode === "lightning";

  // Accumulator for the current open assistant turn — flushed on each tool
  // call or on the final answer. Each entry is one paragraph of prose.
  let reasoningBuffer: string[] = [];

  // Walk steps in their declared `index` order. We don't trust insertion
  // order — the schema defines `index` as the canonical sequence.
  const steps = [...trace.steps].sort((a, b) => a.index - b.index);

  let toolCallCounter = 0;

  const flushAssistant = (
    finalContent: string,
    toolCall?: ChatToolCall,
  ): void => {
    const assistant: AssistantMessage = {
      role: "assistant",
      content: finalContent,
    };
    if (!stripReasoning && reasoningBuffer.length > 0) {
      assistant.reasoning_content = reasoningBuffer.join("\n\n");
    }
    if (toolCall) {
      assistant.tool_calls = [toolCall];
    }
    messages.push(assistant);
    reasoningBuffer = [];
  };

  for (const step of steps) {
    switch (step.kind) {
      case "think":
      case "decide": {
        // Both kinds collapse to plain prose. `decide` is a UI-only annotation
        // for human-authored traces; the trained model never emits it.
        if (stripReasoning) break;
        const text = step.text?.trim();
        if (text) reasoningBuffer.push(text);
        break;
      }
      case "recall": {
        // First-class memory_recall tool call. Replaces v1 prefix syntax.
        const id = `tc-${toolCallCounter++}`;
        const tc: ChatToolCall = {
          id,
          type: "function",
          function: {
            name: TOOL_NAMES.memoryRecall,
            // `arguments` carries the recall query (empty when the schema
            // step doesn't have one — which is fine, the tool returns the
            // ids regardless and the model reads them from the result).
            arguments: jsonStringify({ query: step.text ?? "" }),
          },
        };
        flushAssistant("", tc);
        messages.push({
          role: "tool",
          tool_call_id: id,
          name: TOOL_NAMES.memoryRecall,
          content: jsonStringify({
            claims: step.recalledClaims ?? [],
            episodes: step.recalledEpisodes ?? [],
          }),
        });
        break;
      }
      case "retrieve":
      case "verify":
      case "tool-call": {
        const id = `tc-${toolCallCounter++}`;
        const name = step.tool ?? canonicalToolName(step.kind);
        const tc: ChatToolCall = {
          id,
          type: "function",
          function: {
            name,
            arguments: jsonStringify(step.toolInput),
          },
        };
        flushAssistant("", tc);
        messages.push({
          role: "tool",
          tool_call_id: id,
          name,
          content: jsonStringify(step.toolOutput),
        });
        break;
      }
      case "answer": {
        flushAssistant(step.text ?? "");
        break;
      }
    }
  }

  // Trailing reasoning without a closing answer — emit the assistant turn
  // anyway so the round-trip recovers it. Empty content + reasoning is a
  // valid intermediate state for streamed generations.
  if (reasoningBuffer.length > 0) {
    flushAssistant("");
  }

  return messages;
}

function canonicalToolName(
  kind: "retrieve" | "verify" | "tool-call",
): string {
  if (kind === "retrieve") return TOOL_NAMES.retrieve;
  if (kind === "verify") return TOOL_NAMES.verify;
  return "tool";
}

/* ─────────────────────────────────────────────────────────────
 *  Chat messages  →  ThoughtTrace
 * ──────────────────────────────────────────────────────────── */

/**
 * Reverse of `traceToChatML`. Skips the system + user messages — the trace
 * begins at the first assistant turn.
 *
 * Round-trip guarantee
 * ────────────────────
 * For every trace produced by Veritas-R1 (no `decide` kind, no `confidence`),
 *      `chatMLToTrace(traceToChatML(T)) == T`
 * modulo step-index renumbering (always 0..N on output).
 *
 * For human-authored traces with `decide` / `confidence`, those fields are
 * lost — see the file header.
 */
export function chatMLToTrace(messages: ChatMessage[]): ThoughtTrace {
  const steps: ThoughtStep[] = [];
  let index = 0;

  // Index tool-result messages by their `tool_call_id` for O(1) pairing.
  const toolResultByCallId = new Map<string, ToolMessage>();
  for (const m of messages) {
    if (m.role === "tool") toolResultByCallId.set(m.tool_call_id, m);
  }

  for (const m of messages) {
    if (m.role !== "assistant") continue;

    // 1. Reasoning prose → one `think` step per double-newline-separated
    //    paragraph. We use blank-line separation (encoded as "\n\n" on the
    //    way out) so multi-paragraph thoughts don't collapse into one blob.
    if (m.reasoning_content) {
      const paragraphs = m.reasoning_content
        .split(/\n\s*\n/)
        .map((p) => p.trim())
        .filter((p) => p.length > 0);
      for (const p of paragraphs) {
        steps.push({ kind: "think", text: p, index });
        index++;
      }
    }

    // 2. Tool calls → step kind based on tool name. `memory_recall` is
    //    decoded back to a `recall` step; everything else maps via name.
    if (m.tool_calls) {
      for (const tc of m.tool_calls) {
        const result = toolResultByCallId.get(tc.id);
        const parsedOutput = result ? jsonParseSafe(result.content) : undefined;

        if (tc.function.name === TOOL_NAMES.memoryRecall) {
          // Special-case decode of memory_recall — claim/episode ids live
          // in the tool result payload, not the call. Recover the query
          // text from the call arguments so `step.text` survives round-trip
          // when one was supplied (otherwise the recall step is anonymous).
          const out = (parsedOutput ?? {}) as {
            claims?: unknown;
            episodes?: unknown;
          };
          const recalledClaims = stringArray(out.claims);
          const recalledEpisodes = stringArray(out.episodes);
          const args = jsonParseSafe(tc.function.arguments);
          let queryText: string | undefined;
          if (args && typeof args === "object") {
            const q = (args as { query?: unknown }).query;
            if (typeof q === "string" && q.length > 0) queryText = q;
          }
          const step: ThoughtStep = { kind: "recall", index };
          if (queryText) step.text = queryText;
          if (recalledClaims.length) step.recalledClaims = recalledClaims;
          if (recalledEpisodes.length) step.recalledEpisodes = recalledEpisodes;
          steps.push(step);
          index++;
          continue;
        }

        const kind = canonicaliseToolKind(tc.function.name);
        const step: ThoughtStep = {
          kind,
          index,
          tool: tc.function.name,
          toolInput: jsonParseSafe(tc.function.arguments),
          toolOutput: parsedOutput,
        };
        steps.push(step);
        index++;
      }
    }

    // 3. Final answer — only if non-empty. Empty content on a tool-bearing
    //    turn is structural, not a real answer step.
    if (m.content && m.content.length > 0) {
      steps.push({ kind: "answer", text: m.content, index });
      index++;
    }
  }

  return { steps };
}

/**
 * Map a tool name back to the schema's three tool-bearing kinds. Veritas-R1's
 * default tool surface uses `retrieve` / `verify_citation` / generic; other
 * names fall back to the catch-all `tool-call` kind so the model can call
 * arbitrary tools defined per-deployment without breaking the parser.
 */
function canonicaliseToolKind(
  name: string,
): "retrieve" | "verify" | "tool-call" {
  if (name === TOOL_NAMES.retrieve) return "retrieve";
  if (name === TOOL_NAMES.verify) return "verify";
  // Heuristic for retrieval-flavoured tool names — keeps the schema kind
  // faithful even when the model emits a more specific name.
  if (/verify|doi/i.test(name)) return "verify";
  if (/retrieve|search|crossref|openalex|arxiv|pubmed/i.test(name)) {
    return "retrieve";
  }
  return "tool-call";
}

/* ─────────────────────────────────────────────────────────────
 *  Tiny JSON helpers — never throw; always emit a deterministic
 *  string. Determinism matters because identical traces must
 *  produce identical message arrays for dataset dedup at training.
 * ──────────────────────────────────────────────────────────── */

function jsonStringify(value: unknown): string {
  if (value === undefined) return "{}";
  try {
    return JSON.stringify(value);
  } catch {
    return "{}";
  }
}

function jsonParseSafe(s: string): unknown {
  if (!s) return undefined;
  try {
    return JSON.parse(s);
  } catch {
    return s; // surface the raw string rather than silently dropping data
  }
}

function stringArray(v: unknown): string[] {
  if (!Array.isArray(v)) return [];
  const out: string[] = [];
  for (const x of v) if (typeof x === "string") out.push(x);
  return out;
}
